crates/egui/src/widgets/plot/items/mod.rs

```diff
@@ -1569,6 +1569,421 @@ impl PlotItem for BoxPlot {
     }
 }
 
+// ----------------------------------------------------------------------------
+// HEATMAP section
+
+#[derive(Debug)]
+pub enum HeatmapErr {
+    ZeroColumns,
+    ZeroRows,
+    BadLength,
+}
+
+impl std::fmt::Display for HeatmapErr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::ZeroColumns => write!(f, "number of columns must not be zero"),
+            Self::ZeroRows => write!(f, "number of rows must not be zero"),
+            Self::BadLength => write!(
+                f,
+                "length of value vector is not divisible by number of columns"
+            ),
+        }
+    }
+}
+
+impl std::error::Error for HeatmapErr {}
+
+/// A heatmap.
+pub struct Heatmap<const RESOLUTION: usize> {
+    /// occupied space in absolute plot coordinates
+    pos: PlotPoint,
+    /// values to plot
+    pub(super) values: Vec<f64>,
+    /// number of columns in heatmap
+    cols: usize,
+    /// number of rows in heatmap
+    rows: usize,
+    /// minimum value in colormap.
+    /// Everything smaller will be mapped to palette[0]
+    min: f64,
```

Comment:
I find readability increases a lot by adding newlines before docstrings:

```suggestion
    pos: PlotPoint,
    
    /// values to plot
    pub(super) values: Vec<f64>,
    
    /// number of columns in heatmap
    cols: usize,
    
    /// number of rows in heatmap
    rows: usize,
    
    /// minimum value in colormap.
    /// Everything smaller will be mapped to palette[0]
    min: f64,
```